vms/platformvm/state/state.go

```diff
@@ -1785,20 +1786,80 @@ func (s *state) writeBlsKeyDiffs(height uint64, blsKeyDiffs map[ids.NodeID]*bls.
 	return nil
 }
 
+func (s *state) updateValidatorSet(
+	updateValidators bool,
+	valSetDiff map[subnetNodeKey]validatorStatusPair,
+	weightDiffs map[subnetNodeKey]*ValidatorWeightDiff,
+) error {
+	if !updateValidators {
+		return nil
+	}
+
+	for key, weightDiff := range weightDiffs {
+		var (
+			subnetID = key.subnetID
+			nodeID   = key.nodeID
+			val      = valSetDiff[key]
+			err      error
+		)
+
+		if weightDiff.Amount == 0 {
+			// No weight change to record; go to next validator.
+			continue
+		}
+
+		if weightDiff.Decrease {
+			err = s.cfg.Validators.RemoveWeight(subnetID, nodeID, weightDiff.Amount)
+		} else {
+			if val.status == added {
+				staker := val.validator
+				err = s.cfg.Validators.AddStaker(
+					subnetID,
+					nodeID,
+					staker.PublicKey,
+					staker.TxID,
+					weightDiff.Amount,
+				)
+			} else {
+				err = s.cfg.Validators.AddWeight(subnetID, nodeID, weightDiff.Amount)
+			}
+		}
+		if err != nil {
+			return fmt.Errorf("failed to update validator weight: %w", err)
+		}
+	}
+
+	totalWeight, err := s.cfg.Validators.TotalWeight(constants.PrimaryNetworkID)
+	if err != nil {
+		return fmt.Errorf("failed to get total weight of primary network: %w", err)
+	}
+
+	s.metrics.SetLocalStake(s.cfg.Validators.GetWeight(constants.PrimaryNetworkID, s.ctx.NodeID))
+	s.metrics.SetTotalStake(totalWeight)
+	return nil
+}
+
 type subnetNodeKey struct {
 	subnetID ids.ID
 	nodeID   ids.NodeID
 }
 
+type validatorStatusPair struct {
+	validator *Staker
+	status    diffValidatorStatus
+}
+
 type diffs struct {
 	weightDiffs map[subnetNodeKey]*ValidatorWeightDiff
 	blsKeyDiffs map[ids.NodeID]*bls.PublicKey
+	valSetDiff  map[subnetNodeKey]validatorStatusPair
 }
 
 func (s *state) calculateDiffs() (diffs, error) {
 	var (
 		outputWeights = make(map[subnetNodeKey]*ValidatorWeightDiff)
 		outputBlsKey  = make(map[ids.NodeID]*bls.PublicKey)
+		outputValSet  = make(map[subnetNodeKey]validatorStatusPair)
```

Comment:
Should we just create a `diffs` struct here instead, and populate its fields below? Idc much either way but worth considering. Also we should consider pre-allocating these, even if the initial sizes are small (e.g. `8` or `16`). This is probably a hot path

Reply:
Partly done, not sure which sizes to pick.